File: app/src/main/java/com/ubiqsmart/ui/transactions/TransactionsFragment.java
package com.ubiqsmart.ui.transactions;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.Response;
import com.ubiqsmart.R;
import com.ubiqsmart.ui.detail.AddressDetailActivity;
import com.ubiqsmart.repository.data.TransactionDisplay;
import com.ubiqsmart.network.EtherscanAPI;
import com.ubiqsmart.utils.RequestCache;
import com.ubiqsmart.utils.ResponseParser;

import java.io.*;
import java.util.*;

import static android.view.View.GONE;

public class TransactionsFragment extends TransactionsAbstractFragment {

  public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
    View rootView = super.onCreateView(inflater, container, savedInstanceState);
    getSend().setVisibility(GONE);
    getRequest_transaction().setVisibility(GONE);
    getFabmenu().setVisibility(View.GONE);
    return rootView;
  }

  public void update(boolean force) {
    if (getAc() == null) return;
    resetRequestCount();
    getWallets().clear();
    if (getSwipe_refresh_layout2() != null) getSwipe_refresh_layout2().setRefreshing(true);

    try {
      EtherscanAPI.getInstance().getNormalTransactions(getAddress(), new Callback() {
        @Override public void onFailure(Call call, IOException e) {
          if (isAdded()) {
            getAc().runOnUiThread(new Runnable() {
              @Override public void run() {
                onItemsLoadComplete();
                ((AddressDetailActivity) getAc()).snackError(getString(R.string.err_no_con));
              }
            });
          }
        }

        @Override public void onResponse(Call call, Response response) throws IOException {
          String restring = response.body().string();
          if (restring != null && restring.length() > 2) RequestCache.getInstance().put(RequestCache.TYPE_TXS_NORMAL, getAddress(), restring);
          final List<TransactionDisplay> w =
              new ArrayList<TransactionDisplay>(ResponseParser.parseTransactions(restring, "Unnamed Address", getAddress(), TransactionDisplay.NORMAL));
          if (isAdded()) {
            getAc().runOnUiThread(new Runnable() {
              @Override public void run() {
                onComplete(w);
              }
            });
          }
        }
      }, force);
      EtherscanAPI.getInstance().getInternalTransactions(getAddress(), new Callback() {
        @Override public void onFailure(Call call, IOException e) {
          if (isAdded()) {
            getAc().runOnUiThread(new Runnable() {
              @Override public void run() {
                onItemsLoadComplete();
                ((AddressDetailActivity) getAc()).snackError(getString(R.string.err_no_con));
              }
            });
          }
        }

        @Override public void onResponse(Call call, Response response) throws IOException {
          String restring = response.body().string();
          if (restring != null && restring.length() > 2) RequestCache.getInstance().put(RequestCache.TYPE_TXS_INTERNAL, getAddress(), restring);
          final List<TransactionDisplay> w =
              new ArrayList<TransactionDisplay>(ResponseParser.parseTransactions(restring, "Unnamed Address", getAddress(), TransactionDisplay.CONTRACT));
          if (isAdded()) {
            getAc().runOnUiThread(new Runnable() {
              @Override public void run() {
                onComplete(w);
              }
            });
          }
        }
      }, force);
    } catch (IOException e) {
      if (getAc() != null) ((AddressDetailActivity) getAc()).snackError("Can't fetch account balances. No connection?");
      onItemsLoadComplete();
      e.printStackTrace();
    }
  }

  private void onComplete(List<TransactionDisplay> w) {
    addToWallets(w);
    addRequestCount();
    if (getRequestCount() >= 2) {
      onItemsLoadComplete();
      getNothing_found().setVisibility(getWallets().size() == 0 ? View.VISIBLE : View.GONE);
      getWalletAdapter().notifyDataSetChanged();
    }
  }

}

File: app/src/main/java/com/ubiqsmart/app/ui/transactions/TransactionsFragment.kt
package com.ubiqsmart.app.ui.transactions

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.View.GONE
import android.view.ViewGroup
import com.github.salomonbrys.kodein.instance
import com.github.salomonbrys.kodein.with
import com.ubiqsmart.R
import com.ubiqsmart.datasource.api.EtherscanAPI
import com.ubiqsmart.domain.models.TransactionDisplay
import com.ubiqsmart.app.ui.detail.AddressDetailActivity
import com.ubiqsmart.app.utils.AddressNameConverter
import com.ubiqsmart.app.utils.ResponseParser
import com.ubiqsmart.app.utils.WalletStorage
import kotlinx.android.synthetic.main.fragment_transaction.*
import okhttp3.Call
import okhttp3.Callback
import okhttp3.Response
import java.io.IOException
import java.util.*

class TransactionsFragment : TransactionsAbstractFragment() {

  override val walletStorage: WalletStorage by with(this).instance()
  override val addressNameConverter: AddressNameConverter by with(this).instance()
  override val etherscanApi: EtherscanAPI by instance()

  override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
    val rootView = super.onCreateView(inflater, container, savedInstanceState)

    new_transaction.visibility = GONE
    request_transaction?.visibility = GONE
    fab_menu_view.visibility = View.GONE

    return rootView
  }

  override fun update(force: Boolean) {
    if (activity == null) {
      return
    }

    resetRequestCount()
    wallets.clear()

    if (swipe_refresh_layout2 != null) {
      swipe_refresh_layout2!!.isRefreshing = true
    }

    try {
      etherscanApi.getNormalTransactions(address, object : Callback {
        override fun onFailure(call: Call, e: IOException) {
          if (isAdded) {
            activity.runOnUiThread {
              onItemsLoadComplete()
              (activity as AddressDetailActivity).snackError(getString(R.string.err_no_con))
            }
          }
        }

        @Throws(IOException::class)
        override fun onResponse(call: Call, response: Response) {
          val restring = response.body()!!.string()

          val w = ArrayList(ResponseParser.parseTransactions(restring, getString(R.string.unnamed_address), address, TransactionDisplay.NORMAL))
          if (isAdded) {
            activity.runOnUiThread { onComplete(w) }
          }
        }
      })

      etherscanApi.getInternalTransactions(address, object : Callback {
        override fun onFailure(call: Call, e: IOException) {
          if (isAdded) {
            activity.runOnUiThread {
              onItemsLoadComplete()
              (activity as AddressDetailActivity).snackError(getString(R.string.err_no_con))
            }
          }
        }

        @Throws(IOException::class)
        override fun onResponse(call: Call, response: Response) {
          val restring = response.body()!!.string()
          val w = ArrayList(ResponseParser.parseTransactions(restring, getString(R.string.unnamed_address), address, TransactionDisplay.CONTRACT))
          if (isAdded) {
            activity.runOnUiThread { onComplete(w) }
          }
        }
      })
    } catch (e: IOException) {
      if (activity != null) {
        (activity as AddressDetailActivity).snackError(getString(R.string.cant_fetch_accounts))
      }

      onItemsLoadComplete()

      e.printStackTrace()
    }

  }

  private fun onComplete(w: List<TransactionDisplay>) {
    addToWallets(w)
    addRequestCount()
    if (requestCount >= 2) {
      onItemsLoadComplete()
      nothing_found?.visibility = if (wallets.size == 0) View.VISIBLE else View.GONE
      walletAdapter?.notifyDataSetChanged()
    }
  }

}